creator-node/src/snapbackSM/snapbackSM.js

```diff
@@ -1071,58 +1094,81 @@ class SnapbackSM {
+          decisionTree,
+          'issueUpdateReplicaSetOp() Error',
+          { error: e.message }
+        )
         throw new Error(
           'processStateMachineOperation():issueUpdateReplicaSetOp() Error'
         )
       }
-
-      decisionTree.push({
-        stage: 'END processStateMachineOperation()',
-        vals: {
-          currentModuloSlice: this.currentModuloSlice,
-          moduloBase: this.moduloBase,
-          numSyncRequestsEnqueued,
-          numUpdateReplicaOpsIssued
-        },
-        time: Date.now()
-      })
-
-      // Log error without throwing - next run will attempt to rectify
     } catch (e) {
-      decisionTree.push({
-        stage: 'processStateMachineOperation Error',
-        vals: e.message,
-        time: Date.now()
-      })
+      this._addToStateMachineQueueDecisionTree(
+        decisionTree,
+        'processStateMachineOperation() Error',
+        { error: e.message }
+      )
     } finally {
-      // Increment and adjust current slice by this.moduloBase
-      this.currentModuloSlice += 1
-      this.currentModuloSlice = this.currentModuloSlice % this.moduloBase
+      this._addToStateMachineQueueDecisionTree(
+        decisionTree,
+        'END processStateMachineOperation()',
+        {
+          currentModuloSlice: this.currentModuloSlice,
+          moduloBase: this.moduloBase
+        }
+      )
 
       // Log decision tree
-      try {
-        this.log(
-          `processStateMachineOperation Decision Tree ${JSON.stringify(
-            decisionTree
-          )}`
-        )
-      } catch (e) {
-        this.logError(
-          `Error printing processStateMachineOperation Decision Tree ${decisionTree}`
-        )
+      this._printStateMachineQueueDecisionTree(decisionTree)
+
+      // Increment and adjust current slice by this.moduloBase
+      this.currentModuloSlice = (this.currentModuloSlice + 1) % this.moduloBase
+
+      // Release lock
+      await redis.lock.removeLock(lockKey)
+    }
+  }
+
+  _addToStateMachineQueueDecisionTree(decisionTree, stage, data = {}) {
+    const obj = { stage, data, time: Date.now() }
+
+    if (decisionTree.length > 0) {
+      // Set duration if both objs have time field
+      const lastObj = decisionTree[decisionTree.length - 1]
+      if (lastObj && lastObj.time) {
+        const duration = obj.time - lastObj.time
+        obj.duration = duration
       }
     }
+    decisionTree.push(obj)
+  }
+
+  _printStateMachineQueueDecisionTree(decisionTree, msg = '') {
+    // Compute and record `fullDuration`
+    if (decisionTree.length > 2) {
+      const startTime = decisionTree[0].time
+      const endTime = decisionTree[decisionTree.length - 1].time
+      const duration = endTime - startTime
+      decisionTree[decisionTree.length - 1].fullDuration = duration
+    }
+    try {
+      this.log(
+        `processStateMachineOperation() Decision Tree${
+          msg ? ` - ${msg} - ` : ''
```

Comment:
is this some special syntax to print key - value?

Reply:
no its just a ternary operator if `msg` is truthy it'll print it else it prints empty string